csrc/ir/utils.h

```diff
@@ -164,20 +164,31 @@ std::vector<int> normalizeOld2New(
     const std::unordered_map<int, int>& old2new_in,
     size_t ndims);
 
-// Replace all uses of reference with substitute in expr. Return the Expr.
-// Warning: Invalidates provided Expr.
-// Warning: Removes connection of reference through provided Expr.
-// Warning: Creates new Expr connecting substitue.
-// Reference is found through direct pointer comparison.
-TORCH_CUDA_CU_API Expr* replaceValInExpr(
+//! Replaces reference Val with substitute in all Expr inputs and attributes.
+//! Warning: Invalidates provided Expr.
+//! Warning: Removes connection of reference through provided Expr.
+//! Warning: Creates new Expr defining substitute.
+TORCH_CUDA_CU_API Expr* replaceValInExprInputs(
```

Comment:
How is this different from https://github.com/NVIDIA/Fuser/blob/178ed873baed1ad9b8e9bc2070551f0bb1f23f1f/csrc/device_lower/utils.h#L168-L176? Can we unify them?

Reply:
Yeah that is the kernel version of this. The `ReplaceExprInput` class is a `kir::ExprMutator` that behaves pretty  similarly to what `OptOutMutator` does now, but since it is expression tree-based it keeps track of scopes, positions, and predicates of replaced expressions. Currently it has a separate `handle()` override for each handled expression type; we could probably just compose this class with `OptOutMutator` and call `OptOutMutator::mutate` to create replacements. That might simplify the `ReplaceExprInput` class a bit.